sympy/functions/elementary/piecewise.py

```diff
@@ -664,3 +993,210 @@ def piecewise_fold(expr):
+
+
+def _pair(A, B, union=False, dict=None, abi=False):
+    """Return the intervals on the real number line that are covered
+    by the two intervals, A = [a, b] and B = [c, d]. If union is
+    true then a simple list of intervals is returned; if dict is
+    true (default) then the intervals are keyed to 0 if they are
+    covered by A, 1 if they are covered by B and -1 if they are covered
+    by neither, with higher precedence given to A. If abi is true,
+    return an ordered set of tuples (a, b, i) where
+    a is the lower bound, b is the upper bound and i indicates which
+    interval covers that range: 0 for A, 1 for B and -1 for neither.
+
+    Examples
+    ========
+
+    >>> from sympy import Tuple
+    >>> from sympy.functions.elementary.piecewise import _pair
+    >>> from sympy.abc import x
+
+    For the following segments,
+
+           1---2
+                      4---5
+
+    5 regions are identified:
+
+        -oo -> 1 -> 2 -> 4 -> 5 -> oo
+
+    >>> A = (1, 2); B = (4, 5)
+    >>> _pair(A, B, abi=True)
+    [(-oo, 1, -1), (1, 2, 0), (2, 4, -1), (4, 5, 1), (5, oo, -1)]
+    >>> _pair(A, B, union=True)
+    [(1, 2), (4, 5)]
+    >>> _pair(A, B, dict=True)  # default
+    {-1: [(-oo, 1), (2, 4), (5, oo)], 0: [(1, 2)], 1: [(4, 5)]}
+
+    When one or more of the boundaries are symbolic, the results
+    become more complex and the boundaries are given in terms of
+    Min and Max functions:
+
+    >>> _pair((x, 2), (4, 5))
+    {-1: [(-oo, Min(4, x)), (Min(5, x), x), (2, 4), (5, oo)],
+    0: [(x, 2)],
+    1: [(Min(4, x), Min(5, x)), (4, 5)]}
+
+    When an actual value is given for x, there may be redundant
+    intervals and intervals with no width:
+
+    >>> last = _
+    >>> def replace(x, y):
+    ...     rv = {}
+    ...     for k, v in last.items():
+    ...         rv[k] = [Tuple(*i).subs(x, y) for i in v]
+    ...     return rv
+    ...
+    >>> for k, v in sorted(replace(x, 1).items()):
+    ...     k, v
+    (-1, [(-oo, 1), (1, 1), (2, 4), (5, oo)])
+    (0, [(1, 2)])
+    (1, [(1, 1), (4, 5)])
+
+    There may also be intervals wherein the right boundary is greater
+    than the left and redundant intervals:
+
+    >>> for k, v in sorted(replace(x, 6).items()):
+    ...     (k, v)
+    (-1, [(-oo, 4), (5, 6), (2, 4), (5, oo)])
+    (0, [(6, 2)])
+    (1, [(4, 5), (4, 5)])
+    """
+    if not any(i for i in (abi, dict, union)):
+        dict = True
+    if [bool(i) for i in (abi, dict, union)].count(True) != 1:
+        raise ValueError('only 1 of abi, dict and union may be True')
+
+    a, b = A
+    c, d = B
+    oo = S.Infinity
+    # the real line is broken into pieces which are compactly
+    # represented as -oo, i1, H1|L2, i2, H2|L3, i3, ..., oo
+    # where i1, i2, ...., refer to the interval A (0), B (1) or
+    # neither (-1) as covering that interval and Hi|Lj means
+    # either the hi boundary of the ith interval or the lo
+    # boundary of the jth interval
+    if (b <= c) is S.true:
```

Comment:
This does not seem to work as expected:
```
>>> (1 <= 2) is S.true
False
```

Reply:
If you use `is` to compare against S.true, you had better make sure your arguments are sympified. It's more robust to just use `== S.true`.